On chain wiki (#196)
gear-tech/wiki: docs/examples/onchain-nft.md

---
sidebar_label: OnChain NFT
sidebar_position: 16
---

# On-chain gNFT assets

### Introduction
On the info related to gNFT consider reading: [gNFT-721](https://wiki.gear-tech.io/examples/gnft-721).

When the owner of a given token ID wishes to transfer it to another user, it is easy to verify ownership and reassign the token to a new owner.

> Mostly NFTs images (or other base resourse) is stored somewhere else (e.g. ipfs) and only the metadata is stored in the contract. Metadata consists of a name, an ID and links to the external resources, where the images is actually stored.

Reviewer: ```suggestion
Mostly NFTs images (or other base resources) are stored somewhere else (e.g. IPFS) and only the metadata is stored in the contract. Metadata consists of a name, an ID and links to the external resources, where the images are actually stored.
```

> But there is another approached introduced here. Sometimes you can store NFTs directly on chain without any external storage. This approach helps you not to lose your nft if there is a problem with the external storage.

Reviewer: ```suggestion
But there is another approach introduced here. Sometimes you can store NFTs directly on chain without any external storage. This approach helps you not to lose your ТАЕ if there is a problem with the external storage.
```

        OnChainNFTAction::Approve { to, token_id } => NFTCore::approve(nft, &to, token_id),
    }
}

#[no_mangle]
pub unsafe extern "C" fn meta_state() -> *mut [i32; 2] {
    let query: OnChainNFTQuery = msg::load().expect("failed to decode input argument");
    let nft = CONTRACT.get_or_insert(OnChainNFT::default());
    match query {
        OnChainNFTQuery::TokenURI { token_id } => {
            let encoded = OnChainNFTCore::token_uri(nft, token_id)
                .expect("Error in reading OnChainNFT contract state");
            gstd::util::to_leak_ptr(encoded)
        }
        OnChainNFTQuery::Base(query) => {
            let encoded =
                NFTMetaState::proc_state(nft, query).expect("Error in reading NFT contract state");
            gstd::util::to_leak_ptr(encoded)
        }
    }
}
```

## Conclusion

Gear provides a reusable [library](https://github.com/gear-dapps/non-fungible-token/tree/master/nft/src) with core functionality for the gNFT protocol. By using object composition, that library can be utilized within a custom NFT contract implementation in order to minimize duplication of community available code.

> A source code of the NFT marketplace provided by Gear is available on GitHub: [on-chain-nft/src](https://github.com/gear-dapps/non-fungible-token/tree/master/on-chain-nft/src).

Reviewer: marketplace?

---
sidebar_label: OnChain NFT

> sidebar_position: 14

Reviewer 2: ```suggestion
sidebar_position: 16
```

---

> # Gear Non-Fungible Token

Reviewer 2: ```suggestion
# On-chain gNFT assets
```

> ### Introduction

Reviewer 2: Pls keep only essential info about gNFT, for the rest - provide a link on existing page: https://wiki.gear-tech.io/examples/gnft-721

Author: Fixed.

When the owner of a given token ID wishes to transfer it to another user, it is easy to verify ownership and reassign the token to a new owner.

Mostly NFTs images (or other base resourse) is stored somewhere else (e.g. ipfs) and only the metadata is stored in the contract. Metadata consists of a name, an ID and links to the external resources, where the images is actually stored.
But there is another approached introduced here. Sometimes you can store NFTs directly on chain without any external storage. This approach helps you not to lose your nft if there is a problem with the external storage.

### Approach
To successfully implement this approach several things are needed. Firstly, when initializing a collection, one should provide all the possible images of all the layers for a collection. Secondly, when minting alongside with a small metadata, one should provide a combination of layers used for a specific NFT. This approach seems quite costly when initializing, but is relatively cheap when it comes to minting.

> ### Default non-fungible-token implementation

Reviewer 2: Show the specific of on-chain NFT, keep only necessary info that needs to understand the difference of common NFT implementation from on-chain NFT.
